src/dataflow/src/source/s3.rs

```diff
@@ -134,26 +136,42 @@ async fn download_objects_task(
                     seen_buckets.insert(msg.bucket.clone(), bi);
                 };
 
-                let update = download_object(
-                    &tx,
-                    &activator,
-                    &client,
-                    msg.bucket.clone(),
-                    msg.key,
-                    compression,
-                )
-                .await;
-
-                if let Some(update) = update {
-                    seen_buckets
-                        .get_mut(&msg.bucket)
-                        .expect("just inserted")
-                        .metrics
-                        .inc(1, update.bytes, update.messages);
-                    if update.sent == Sent::SenderClosed {
-                        rx.close();
+                let mut retries_available = 10;
+                while retries_available > 0 {
```

Comment:
I thiiiink you could reduce the number of nestings here by putting the retry count onto the match expression, like 

``` rust
match (msg, retries_available) {
    (Err(msg), 0) => { break; }
    (Err(msg), _) => ...
    (Ok(msg), _) => ...
```

Ought to make the loop-breaking situation less complicated, too (:

Reply:
Merging the retry-handling with the `'outer' loop is problematic because we need to consider the same message multiple times. This means that its `BucketInfo` needs to accommodate seeing the same key multiple times, and we need to be able to poll an existing key, or `await` a new one. I tried the following, but I'm not sure how to make it work with async:

```
    let mut pending_message = None;

    'outer: while let Some(msg) = pending_message.unwrap_or_else(|| rx.recv().await) {
```

This doesn't work because the closure can't be async for obvious reasons. Is there a work-around?

Reply:
Oh well, converting the `unwrap_or_else` to an explicit `if` statement solves the problem. So this is certainly doable!

Reply:
Hm, I think your objection is reasonable re. having to try and consume the same message multiple times - I hadn't noticed that. If this makes the code more complicated (especially with re-use of data), I'd say the nested loop is fine!

Reply:
Played with combining the retry loop with the outer match. It works, but readability is not improved. Thus I prefer keeping it as it currently is.